Source/JavaScriptCore/runtime/RegExpConstructor.cpp

```diff
@@ -83,18 +86,96 @@ RegExpConstructor::RegExpConstructor(VM& vm, Structure* structure)
 {
 }
 
-void RegExpConstructor::finishCreation(VM& vm, RegExpPrototype* regExpPrototype)
+void RegExpConstructor::finishCreation(VM& vm, JSGlobalObject* globalObject, RegExpPrototype* regExpPrototype)
 {
     Base::finishCreation(vm, 2, vm.propertyNames->RegExp.string(), PropertyAdditionMode::WithoutStructureTransition);
     ASSERT(inherits(info()));
 
     putDirectWithoutTransition(vm, vm.propertyNames->prototype, regExpPrototype, PropertyAttribute::DontEnum | PropertyAttribute::DontDelete | PropertyAttribute::ReadOnly);
 
-    JSGlobalObject* globalObject = regExpPrototype->globalObject();
+    if (Options::useRegExpEscape())
+        JSC_NATIVE_FUNCTION_WITHOUT_TRANSITION("escape"_s, regExpConstructorEscape, static_cast<unsigned>(PropertyAttribute::DontEnum), 1, ImplementationVisibility::Public);
+
     GetterSetter* speciesGetterSetter = GetterSetter::create(vm, globalObject, JSFunction::create(vm, globalObject, 0, "get [Symbol.species]"_s, globalFuncSpeciesGetter, ImplementationVisibility::Public, SpeciesGetterIntrinsic), nullptr);
     putDirectNonIndexAccessorWithoutTransition(vm, vm.propertyNames->speciesSymbol, speciesGetterSetter, PropertyAttribute::Accessor | PropertyAttribute::ReadOnly | PropertyAttribute::DontEnum);
 }
 
+JSC_DEFINE_HOST_FUNCTION(regExpConstructorEscape, (JSGlobalObject* globalObject, CallFrame* callFrame))
+{
+    VM& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
+    JSValue argValue = callFrame->argument(0);
+    if (!argValue.isString())
+        return throwVMTypeError(globalObject, scope, "RegExp.escape requires a string"_s);
+
+    JSString* argString = asString(argValue);
+    ASSERT(argString);
+
+    const String& string = argString->value(globalObject);
+    RETURN_IF_EXCEPTION(scope, { });
+
+    StringBuilder builder;
+    builder.reserveCapacity(string.length());
+
+    for (unsigned i = 0; i < string.length();) {
+        char32_t codePoint;
+        if (string.is8Bit())
+            codePoint = string.characters8()[i++];
+        else
+            U16_NEXT(string.characters16(), i, string.length(), codePoint);
+
+        if (builder.isEmpty() && isASCIIAlphanumeric(codePoint)) {
+            builder.append('\\', 'x', toStringWithRadix(codePoint, 16));
+            continue;
+        }
+
+        if (isASCII(codePoint) && strchr("^$\\.*+?()[]{}|/", codePoint)) {
```

Comment:
`strchr` seems rather unusual; I would expect `.contains` here.

Reply:
this is using a raw string though, not a `String`.  do we have a `static` version of `contains` that takes a `const char*`?  or are you suggesting that i create a `String`/`StringView`?

FWIW ive seen this done in a few other places (e.g. `JSC::Yarr::Parser`, `Inspector::ContentSearchUtilities`, etc.)

Reply:
I was suggesting that it wouldn't just be a raw string, yeah—I think `AtomString` (i.e. `_s` would be appropriate?).
That said, I'd just been grepping in `runtime/`, so I missed the two cases you mentioned, which are indeed super similar.

Reply:
While it's acceptable either way based on the precedents you gave, I think it would be better to avoid `const char*`—ChrisD just landed a patch about exactly this a few hours ago: https://github.com/WebKit/WebKit/pull/27549

Reply:
Ah crap, I said `AtomString` when I meant `ASCIILiteral`. 🤦 I think maybe `StringView(ASCIILiteral)` would be more appropriate here, as we have in [IntlNumberFormat.cpp](https://github.com/WebKit/WebKit/blob/main/Source/JavaScriptCore/runtime/IntlNumberFormat.cpp#L212).